### PR TITLE
Put crate metadata first in the rlib

### LINKED ISSUE #93806: Put crate metadata first in rlibs whenever possible
In the past rlib metadata had to be the last archive member as the linker stops at the first non-object member. The rlib metadata has somewhat recently been changed to be wrapped in an object file on most platforms, which allows it to be anywhere in the rlib. Putting it first would allow for quicker metadata reading as it would no longer be necessary to read the header of every archive member to find the crate metadata, only the first member header.

### PR DESCRIPTION
**Pull request by bjorn3**
This should make metadata lookup faster

Fixes https://github.com/rust-lang/rust/issues/93806

### CODE REVIEW

**Comment by bjorn3** [comment]
It seems like this slightly reduces cycle count overall, a bigger inprovement in fault count and a significant up to 6% reduction in max-rss. Wall time is too noisy to notice any effect.

**Comment by bjorn3** [comment]
https://github.com/rust-lang/rust/pull/93864 has landed.

@rustbot ready

File: compiler/rustc_codegen_ssa/src/back/link.rs [hunk 1]
```diff
@@ -266,6 +266,27 @@ fn link_rlib<'a, B: ArchiveBuilder<'a>>(
 
     let mut ab = <B as ArchiveBuilder>::new(sess, out_filename, None);
 
+    let metadata = match flavor {
```
Comment (nagisa) [inline compiler/rustc_codegen_ssa/src/back/link.rs:269]: Nit:

```suggestion
    let trailing_metadata = match flavor {
```

perhaps? plain `metadata` seems overly generic for the somewhat nontrivial logic happening here.

File: compiler/rustc_codegen_ssa/src/back/metadata.rs [hunk 2]
```diff
@@ -199,7 +199,7 @@ fn create_object_file(sess: &Session) -> Option<write::Object<'static>> {
 // * ELF - All other targets are similar to Windows in that there's a
 //   `SHF_EXCLUDE` flag we can set on sections in an object file to get
 //   automatically removed from the final output.
-pub fn create_rmeta_file(sess: &Session, metadata: &[u8]) -> Vec<u8> {
+pub fn create_rmeta_file(sess: &Session, metadata: &[u8]) -> (Vec<u8>, bool) {
```
Comment (nagisa) [inline compiler/rustc_codegen_ssa/src/back/metadata.rs:202]: Nit: would it make sense to replace this `bool` with an enum? It isn't entirely obvious at a glance what this represents and also it seems like it'd be quite straightforward to accidentally ignore it, even though the data is quite different depending on the branch taken.

Comment (bjorn3) [reply compiler/rustc_codegen_ssa/src/back/metadata.rs]: Done

**Comment by nagisa** [comment]
r=me, probably want to squash away the fix review comments commit.

**Comment by bjorn3** [comment]
Unlike the previous perf run this slightly increases instruction count. It does still result in a decent max-rss reduction.

**Pull request merged**